scripts/gha/integration_testing/automated_testapp/AutomatedTestRunner.cs

```diff
@@ -244,4 +215,190 @@ public class MainThreadDispatcher : MonoBehaviour {
+      if (currentAttempt < maxAttempts) {
+        testExceptions.Clear();
+        TookMultipleAttempts = true;
+        currentAttempt++;
+        StartTestInternal();
+        return true;
+      }
+      return false;
+    }
+
+    public void Update() {
+      // Check the Task to determine if it is finished
+      if (Status == AutomatedTestStatus.Running) {
+        if (Result == null || Result.IsCompleted || Result.IsCanceled || Result.IsFaulted) {
+          // Determine if the test failed
+          bool failed = (Result == null || Result.IsCanceled || Result.IsFaulted);
+          // If it gathered exceptions, and they weren't expected, fail
+          failed |= (postTestIgnoredFailureCheckEnabled && testExceptions.Count > 0);
+
+          // If it failed, attempt to retry it
+          if (failed) {
+            if (TryRetryTest()) {
+              return;
+            }
+          }
+          // Otherwise, finish the test
+          Status = failed ? AutomatedTestStatus.Failed : AutomatedTestStatus.Succeeded;
+        } else if (Time.time - startTime > Runner.TestTimeoutSeconds) {
+          // The test has timed out
+          // Try to restart it
+          if (TryRetryTest()) {
+            return;
+          }
+          // Otherwise, finish the test
+          timeOutTime = Time.time - startTime;
+          Status = AutomatedTestStatus.TimedOut;
+        }
+      }
+    }
+
+    public void LogResult() {
+      if (Status == AutomatedTestStatus.Succeeded) {
+        LogFunc("Test " + Description + " passed.");
+      } else if (Status == AutomatedTestStatus.Failed) {
+        if (Result == null) {
+          LogFunc(String.Format("Test {0} failed.\n\n" +
+                                "No task was returned by the test.",
+                                Description));
+        } else if (Result.IsFaulted) {
+          LogFunc(String.Format("Test {0} failed!\n\n" +
+                                "Exception message: {1}",
+                                Description,
+                                Result.Exception.Flatten().ToString()));
+        } else {
+          // Assume exceptions were thrown, causing the failure
+          LogFunc(String.Format("Test {0} failed with {1} exception(s).\n\n",
+                                Description, testExceptions.Count.ToString()));
+          for (int i = 0; i < testExceptions.Count; ++i) {
+            LogFunc(String.Format("Exception message ({0}): {1}\n\n", (i + 1).ToString(),
+                                  testExceptions[i].ToString()));
+          }
+        }
+      } else if (Status == AutomatedTestStatus.TimedOut) {
+        LogFunc("Test " + Description + " timed out, elapsed time: " + timeOutTime);
+      }
+    }
+
+    public void CollectWarnings(List<string> warnings) {
+      // If multiple attempts were taked, add a warning
+      if (TookMultipleAttempts) {
+        warnings.Add(String.Format("Test {0} took {1} of {2} attempts to pass.",
```

Comment:
Should we add the word Flake somewhere in this message?

Reply:
Done